code/include/playfab_c/PlayFabClientApi_c.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include <httpClient/pal.h>
+#include <httpClient/async.h>
+#include "PlayFabClientDataModels_c.h"
+#include "PlayFabLogin_c.h"
+
+/// <summary>
+/// Retrieves a player's profile.
+/// </summary>
+/// <param name="user">PlayFabUserHandle used to authenticate.</param>
+/// <param name="request">Populated request object.</param>
+/// <param name="async">XAsyncBlock for the async operation.</param>
+/// <returns>Result code for this API operation.</returns>
+/// <remarks>
+/// If successful, call <see cref="PlayFabGetPlayerProfileGetResult"/> to get the result.
+/// </remarks>
+HRESULT PlayFabGetPlayerProfileAsync(
+    _In_ PlayFabUserHandle user,
+    _In_ const PlayFabGetPlayerProfileRequest* request,
+    _In_ XAsyncBlock* async
+) noexcept;
+
+/// <summary>
+/// Get the async result of a successful PlayFabGetPlayerProfileAsync call.
+/// </summary>
+/// <param name="async">XAsyncBlock for the async operation.</param>
+/// <param name="handle">Opaque result handle.</param>
+/// <param name="result">Pointer to the result object.</param>
+/// <returns>Result code for this API operation.</returns>
+/// <remarks>
+/// The lifetime of the result object is tied to the result handle. When the result is no longer needed, call
+/// PlayFabResultCloseHandle to release the result object.
+/// </remarks>
+HRESULT PlayFabGetPlayerProfileGetResult(
+    _In_ XAsyncBlock* async,
+    _Out_ PlayFabResultHandle* handle,
+    _Out_ PlayFabPlayerProfileResult** result

```

Comment:
general curiosity question.  sometimes the APIs fill in results into a caller provided buffer, and other times a result pointer if given with a handle.  what determines the difference in patterns?

Reply:
It depends entirely on the result object.  I deliberately chose 3 APIs that illustrate the different patterns I am planning to use:

1. For complex, deeply nested result objects (or result objects that contain dynamically sized lists/maps) PlayFab SDK will own the result and handle the title a handle to it.  Serializing these objects into a continuous byte buffer (as is required by XAsync) is messy and I wanted to avoid that.  Note that the memory for the result object will still be allocated using the title's memory hooks (though this piece hasn't been written yet).  PlayFabGetPlayerProfileAsync shows this pattern.

2. For result objects that are easily serializable into a buffer but are not statically sized (i.e. they have dynamic length string fields), we will ask the title for a buffer and serialize the result into that buffer.  PlayFabCreateSharedGroupAsync shows this pattern.

3. For result objects that are statically sized, we will just populate a result object owned by the title. PlayFabGetTimeAsync shows this pattern.

---
In reply to: [595568404](https://github.com/jplafonta/XPlatCppSdk/pull/2#discussion_r595568404) [](ancestors = 595568404)